File: c/54.c
/***********************************************************************
 * Project Euler (https://serope.com/github/euler.html)
 * Problem 54
 **********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include "euler.h"

#define IS_RANK(x)	x>=1 && x<=14
#define IS_SUIT(x)	x==CLUBS || x==DIAMONDS || x==HEARTS || x==SPADES

struct hand     new_hand();
int             char_to_poker_int(char c);
int             get_hand_type(struct hand h);
int             get_highest_card(struct hand h, int hand_type);
bool            beats(struct hand player1, struct hand player2);
void            print_hand(struct hand h);


enum {
	//Royal ranks
	JACK=11, QUEEN, KING, ACE,
	
	//Hand types, in order of strength
	HIGHCARD, PAIR, TWOPAIR, TRIPS, STRAIGHT, FLUSH, FULLHOUSE, QUADS, STRAIGHTFLUSH, ROYALFLUSH,
	
	//Suits
	CLUBS, DIAMONDS, HEARTS, SPADES
};


struct hand {
	int* ranks;
	int* suits;
};






int main() {
	/*******************************************************************
	 * 1. Open the file
	 ******************************************************************/
	const char* filename = "54_poker.txt";
	FILE* txtfile = fopen(filename, "r");
	if (!txtfile) {
		printf("Unable to open %s \n", filename);
		return 1;
	}
	
	
	
	/*******************************************************************
	 * 2. Convert the contents of the textfile into hands and evaluate
	 *    them
	 ******************************************************************/
	int seek;
	int wins = 0;
	int current_game = 1;
	
	struct hand player1 = new_hand();
	struct hand player2 = new_hand();
	int ranks_len = 0;
	int suits_len = 0;
	int current_player = 1;
	
	//For every character 'x' in the textfile...
	while ((seek=getc(txtfile)) != EOF) {
		int x = char_to_poker_int(seek);
		
		if (x==-1)
			continue;
		
		//Add 'x' to player 1's hand
		if (current_player==1) {
			if (IS_RANK(x))
				player1.ranks = append(player1.ranks, x, &ranks_len);
				
			else {
				player1.suits = append(player1.suits, x, &suits_len);
				
				//After 5 cards have been added to player 1's hand,
				//print the hand, reset the length values and switch
				//to player 2
				if (suits_len==5) {
					printf("%d. \t", current_game);
					current_game++;
					
					print_hand(player1);
					printf("    ");
					
					ranks_len = 0;
					suits_len = 0;
					current_player = 2;
				}
			}
		}
		
		//Add 'x' to player 2's hand
		else if (current_player==2) {
			if (IS_RANK(x))
				player2.ranks = append(player2.ranks, x, &ranks_len);
			
			else {
				player2.suits = append(player2.suits, x, &suits_len);
				
				//After 5 cards have been added to player 2's hand,
				//both players' hands are pit against each other.
				//Then, free everything and start over.
				if (suits_len==5) {
					print_hand(player2);
					
					if (beats(player1, player2)) {
						wins++;
						printf("    %d win", wins);
						if (wins>1)
							printf("s");
					}
					printf("\n");
					
					free(player1.ranks);
					free(player1.suits);
					player1 = new_hand();
					
					free(player2.ranks);
					free(player2.suits);
					player2 = new_hand();
					
					ranks_len = 0;
					suits_len = 0;
					current_player = 1;
				}
			}
		}
	}


	/*******************************************************************
	 * 3. Print, cleanup, and exit
	 ******************************************************************/
	printf("\nPlayer 1 won %d games! \n", wins);
	fclose(txtfile);
	return 0;
}





void print_hand(struct hand h) {
	for (int c=0; c<5; c++) {
		int rank = h.ranks[c];
		int suit = h.suits[c];
		
		switch (rank) {
			case 2:
			case 3:
			case 4:
			case 5:
			case 6:
			case 7:
			case 8:
			case 9:			printf("%d", rank); break;
			case 10:		printf("T"); break;
			case JACK:		printf("J"); break;
			case QUEEN:		printf("Q"); break;
			case KING:		printf("K"); break;
			case ACE:		printf("A"); break;
		}
		
		switch (suit) {
			case CLUBS:		printf("C"); break;
			case DIAMONDS:	printf("D"); break;
			case HEARTS:	printf("H"); break;
			case SPADES:	printf("S"); break;
		}
		
		printf(" ");
	}
}



struct hand new_hand() {
	struct hand h;
	h.ranks = NULL;
	h.suits = NULL;
	return h;
}



int get_hand_type(struct hand h) {
	//1. Bubble sort the hand
	for (int i=0; i<5; i++)
		for (int j=0; j<4; j++)
			if (h.ranks[j]>h.ranks[j+1]) {
				int placeholder	= h.ranks[j];
				h.ranks[j]		= h.ranks[j+1];
				h.ranks[j+1]	= placeholder;
			}
	
	
	//2. Check flush status
	bool is_flush = false;
	if (h.suits[0]==h.suits[1] && h.suits[1]==h.suits[2] && h.suits[2]==h.suits[3] && h.suits[3]==h.suits[4])
		is_flush = true;
	
	
	//3. Check straight status
	bool is_straight = false;
	if (h.ranks[0]==h.ranks[1]-1 && h.ranks[1]==h.ranks[2]-1 && h.ranks[2]==h.ranks[3]-1 && h.ranks[3]==h.ranks[4]-1)
		is_straight = true;
	
	
	//4. Check royal flush status
	bool is_royal_flush = false;
	if (is_flush && is_straight && h.ranks[4]==ACE)
		is_royal_flush = true;
	
	
	//5. Return flush, straight, straight flush, or royal flush
	if (is_royal_flush)
		return ROYALFLUSH;
	else if (is_flush && is_straight)
		return STRAIGHTFLUSH;
	else if (is_flush)
		return FLUSH;
	else if (is_straight)
		return STRAIGHT;
	
	
	//6. Count how many distinct card ranks are in the hand
	int* distinct = NULL;
	int distinct_len = 0;
	
	for (int i=0; i<5; i++)
		if (!contains(distinct, h.ranks[i], distinct_len))
			distinct = append(distinct, h.ranks[i], &distinct_len);
	free(distinct);
	
	//7. Return quads or full house
	if (distinct_len==2) {
		//Quads
		if (h.ranks[0]==h.ranks[1] && h.ranks[1]==h.ranks[2] && h.ranks[2]==h.ranks[3])
			return QUADS;
		else if (h.ranks[1]==h.ranks[2] && h.ranks[2]==h.ranks[3] && h.ranks[3]==h.ranks[4])
			return QUADS;
		
		//Full house
		if (h.ranks[0]==h.ranks[1] && h.ranks[1]==h.ranks[2])
			return FULLHOUSE;
		else if (h.ranks[2]==h.ranks[3] && h.ranks[3]==h.ranks[4])
			return FULLHOUSE;
	}
	
	
	//8. Return trips or two pair
	else if (distinct_len==3) {
		//Trips
		if (h.ranks[0]==h.ranks[1] && h.ranks[1]==h.ranks[2])
			return TRIPS;
		else if (h.ranks[1]==h.ranks[2] && h.ranks[1]==h.ranks[2])
			return TRIPS;
		else if (h.ranks[2]==h.ranks[3] && h.ranks[3]==h.ranks[4])
			return TRIPS;
		
		//Two pair
		else
			return TWOPAIR;
	}
	
	
	//9. Return one pair
	if (distinct_len==4) {
		if (h.ranks[0]==h.ranks[1])
			return PAIR;
		else if (h.ranks[1]==h.ranks[2])
			return PAIR;
		else if (h.ranks[2]==h.ranks[3])
			return PAIR;
		else if (h.ranks[3]==h.ranks[4])
			return PAIR;
	}
	
	
	//10. Return high card
	return HIGHCARD;
}




bool beats(struct hand player1, struct hand player2) {
	int player1_hand_type = get_hand_type(player1);
	int player1_highest_card = get_highest_card(player1, player1_hand_type);
	
	int player2_hand_type = get_hand_type(player2);
	int player2_highest_card = get_highest_card(player2, player2_hand_type);
	
	//If P1's hand type beats P2's
	if (player1_hand_type > player2_hand_type)
		return true;
	
	//If P2's hand type beats P1's
	else if (player1_hand_type < player2_hand_type)
		return false;
	
	//If P1 and P2 have the same hand type
	else if (player1_highest_card > player2_highest_card)
		return true;
		
	return false;
}





int char_to_poker_int(char c) {
	switch (c) {
		case '2': return 2;
		case '3': return 3;
		case '4': return 4;
		case '5': return 5;
		case '6': return 6;
		case '7': return 7;
		case '8': return 8;
		case '9': return 9;
		case 'T': return 10;
		case 'J': return JACK;
		case 'Q': return QUEEN;
		case 'K': return KING;
		case 'A': return ACE;
		case 'C': return CLUBS;
		case 'D': return DIAMONDS;
		case 'H': return HEARTS;
		case 'S': return SPADES;
	}
	return -1;
}








int get_highest_card(struct hand h, int hand_type) {
	//Bubble sort the hand's ranks
	for (int i=0; i<5; i++)
		for (int j=0; j<4; j++)
			if (h.ranks[j]>h.ranks[j+1]) {
				int placeholder	= h.ranks[j];
				h.ranks[j]		= h.ranks[j+1];
				h.ranks[j+1]	= placeholder;
			}
	
	
	//In these hand types, the final card is the highest
	if (hand_type==ROYALFLUSH || hand_type==STRAIGHTFLUSH || hand_type==STRAIGHT || hand_type==FLUSH || hand_type==HIGHCARD)
		return h.ranks[4];


	//Quads
	else if (hand_type==QUADS) {
		if (h.ranks[0]==h.ranks[1])
			return h.ranks[0];
		else
			return h.ranks[1];
	}
	
	//Full house
	else if (hand_type==FULLHOUSE) {
		if (h.ranks[0]==h.ranks[1] && h.ranks[1]==h.ranks[2])
			return h.ranks[0];
		else
			return h.ranks[4];
	}
	
	//Trips
	else if (hand_type==TRIPS) {
		if (h.ranks[0]==h.ranks[1] && h.ranks[1]==h.ranks[2])
			return h.ranks[0];
		else if (h.ranks[1]==h.ranks[2] && h.ranks[2]==h.ranks[3])
			return h.ranks[1];
		else
			return h.ranks[4];
	}
	
	//Two pair (when sorted, the highest card is always in index 3)
	else if (hand_type==TWOPAIR)
		return h.ranks[3];
	
	//Pair
	else if (hand_type==PAIR) {
		if (h.ranks[0]==h.ranks[1])
			return h.ranks[0];
		else if (h.ranks[1]==h.ranks[2])
			return h.ranks[1];
		else if (h.ranks[2]==h.ranks[3])
			return h.ranks[2];
		else
			return h.ranks[3];
	}
	
	return -1;
}

File: go/54.go
// Project Euler
// 54.go
package main

import (
	"fmt"
	"strings"
	"io/ioutil"
)

const (
	// face cards
	A = 14
	K = 13
	Q = 12
	J = 11
	
	// hand types in descending order of strength
	ROYALFLUSH = iota*-1
	STRAIGHTFLUSH
	QUADS
	FULLHOUSE
	FLUSH
	STRAIGHT
	TRIPS
	TWOPAIR
	ONEPAIR
	HIGHCARD
	
	// suits
	C
	D
	H
	S
)

type Card struct {
	rank int
	suit int
}

type Hand []Card



func main() {
	// read each player's hands
	player1, player2, err := readPokerFile("54_poker")
	if err!=nil {
		panic(err)
	}
	
	// count player 1's wins
	wins := 0
	for i:=0; i<len(player1); i++ {
		hand1 := player1[i]
		hand2 := player2[i]
		fmt.Printf("%d \t %v \t %v \t", i, hand1, hand2)
		if hand1.beats(hand2) {
			wins++
			fmt.Print("wins: ", wins)
		}
		fmt.Print("\n")
	}
	fmt.Println(wins)
}


// readPokerFile reads the given file and returns two slices representing
// all of player 1's hands and all of player 2's hands.
func readPokerFile(filename string) ([]Hand, []Hand, error) {
	// read
	file, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, nil, err
	}
	text := strings.Split(string(file), "\n")
	i := len(text)-1
	text = append(text[:i], text[i+1:]...) // remove final line (EOF)
	
	// parse
	var player1, player2 []Hand
	for _, line := range(text) {
		hand1, hand2 := parseShowdown(line)
		player1 = append(player1, hand1)
		player2 = append(player2, hand2)
	}
	return player1, player2, nil
}

// parseShowdown takes a string like "8C TS KC 9H 4S 7D 2S 5D 3S AC" and
// returns two hands.
func parseShowdown(line string) (Hand, Hand) {
	split := strings.Split(line, " ")
	var hand1, hand2 Hand
	for i, str := range(split) {
		card := parseCard(str)
		if i<5 {
			hand1 = append(hand1, card)
		} else {
			hand2 = append(hand2, card)
		}
	}
	return hand1, hand2
}

// parseCard takes a string like "5H" and returns an equivalent Card.
func parseCard(str string) Card {
	rank := parseByte(str[0])
	suit := parseByte(str[1])
	return Card{rank, suit}
}

// parseByte takes a byte/char and returns the appropriate poker value.
func parseByte(b byte) int {
	switch b {
		case '1':	return 1
		case '2':	return 2
		case '3':	return 3
		case '4':	return 4
		case '5':	return 5
		case '6':	return 6
		case '7':	return 7
		case '8':	return 8
		case '9':	return 9
		case 'T':	return 10
		case 'J':	return J
		case 'Q':	return Q
		case 'K':	return K
		case 'A':	return A
		case 'C':	return C
		case 'D':	return D
		case 'H':	return H
		case 'S':	return S
	}
	return -1
}

// intToStr returns the appropriate poker symbol for the given const.
func intToStr(i int) string {
	switch i {
		case 1:		return "1"
		case 2:		return "2"
		case 3:		return "3"
		case 4:		return "4"
		case 5:		return "5"
		case 6:		return "6"
		case 7:		return "7"
		case 8:		return "8"
		case 9:		return "9"
		case 10:	return "T"
		case J:		return "J"
		case Q:		return "Q"
		case K:		return "K"
		case A:		return "A"
		case C:		return "♣"
		case D:		return "♦"
		case H:		return "♥"
		case S:		return "♠"
	}
	return "-"
}

// String returns a printable form of the hand.
func (h Hand) String() string {
	str := ""
	for _, card := range(h) {
		str += intToStr(card.rank) + intToStr(card.suit) + " "
	}
	return str
}

// isFlush returns true if every card in the hand has the same suit.
func (h Hand) isFlush() bool {
	suit := h[0].suit
	for i:=1; i<5; i++ {
		card := h[i]
		if card.suit != suit {
			return false
		}
	}
	return true
}

// sort performs a bubble sort on the hand (slices are reference types,
// so nothing need be returned).
func (h Hand) sort() {
	for i:=0; i<len(h)-1; i++ {
		for j:=i+1; j<len(h); j++ {
			if h[i].rank > h[j].rank {
				tmp := h[i]
				h[i] = h[j]
				h[j] = tmp
			}
		}
	}
}

// isStraight returns true if the hand has consecutive ranks.
func (h Hand) isStraight() bool {
	for i:=0; i<4; i++ {
		cardA := h[i]
		cardB := h[i+1]
		if cardA.rank != cardB.rank-1 {
			return false
		}
	}
	return true
}

// isRoyalFlush returns true if the hand is a royal flush (a straight
// flush whose highest card is A).
func (h Hand) isRoyalFlush() bool {
	return h.isStraight() && h.isFlush() && h[4].rank == A
}

// isStraightFlush returns true if the hand is a straight flush.
func (h Hand) isStraightFlush() bool {
	return h.isStraight() && h.isFlush() && h[4].rank != A
}

// isQuads returns true if the hand is a four-of-a-kind (either the first
// 4 cards or last 4 cards have the same rank).
func (h Hand) isQuads() bool {
	if sameRanks(h[0:4]) || sameRanks(h[1:5]) {
		return true
	}
	return false
}

// isFullHouse returns true if the hand is a full house (it has both a
// three-of-a-kind and a pair).
func (h Hand) isFullHouse() bool {
	if sameRanks(h[0:3]) && sameRanks(h[3:5]) {			// AAABB
		return true
	} else if sameRanks(h[0:2]) && sameRanks(h[2:5]) {	// BBAAA
		return true
	}
	return false
}

// isTrips returns true if the hand is a three-of-a-kind (either the
// first 3 cards, middle 3 cards, or last 3 cards have the same rank).
func (h Hand) isTrips() bool {
	if sameRanks(h[0:3]) || sameRanks(h[1:4]) || sameRanks(h[2:5]) {
		return true
	}
	return false
}

// isTwoPair returns true if the hand is a two pair.
func (h Hand) isTwoPair() bool {
	if sameRanks(h[0:2]) && sameRanks(h[2:4]) {			// AABBc
		return true
	} else if sameRanks(h[0:2]) && sameRanks(h[3:5]) {	// AAcBB
		return true
	} else if sameRanks(h[1:3]) && sameRanks(h[3:5]) {	// cAABB
		return true
	}
	return false
}

// isOnePair returns true if the hand is a one pair.
func (h Hand) isOnePair() bool {
	for i:=0; i<len(h)-1; i++ {
		if h[i].rank == h[i+1].rank {
			return true
		}
	}
	return false
}

// sameRanks returns true if every card in the given hand is of the same
// rank.
func sameRanks(h Hand) bool {
	for i:=0; i<len(h)-1; i++ {
		if h[i].rank != h[i+1].rank {
			return false
		}
	}
	return true
}

// eval evaluates the hand and returns its hand type and highest card.
func (h Hand) eval() (int, Card) {
	h.sort()
	var handType int
	var highest Card
	
	switch {
		case h.isRoyalFlush():
			handType = ROYALFLUSH
			highest = h[4]
		case h.isStraightFlush():
			handType = STRAIGHTFLUSH
			highest = h[4]
		case h.isQuads():
			handType = QUADS
			highest = h[2]
		case h.isFullHouse():
			handType = FULLHOUSE
			highest = h[2]
		case h.isFlush():
			handType = FLUSH
			highest = h[4]
		case h.isStraight():
			handType = STRAIGHT
			highest = h[4]
		case h.isTrips():
			handType = TRIPS
			highest = h[2]
		case h.isTwoPair():
			handType = TWOPAIR
			highest = h[3]
		case h.isOnePair():
			handType = ONEPAIR
			for i:=0; i<len(h)-1; i++ {
				if h[i].rank == h[i+1].rank {
					highest = h[i]
					break
				}
			}
		default:
			handType = HIGHCARD
			highest = max(h)
	}
	return handType, highest
}

// max returns the highest-valued card from a hand.
func max(h Hand) Card {
	m := h[0]
	for _, card := range(h) {
		if card.rank > m.rank {
			m = card
		}
	}
	return m
}


// beats returns true if player 1's hand beats player 2's hand.
func (player1 Hand) beats(player2 Hand) bool {
	player1handType, player1highestCard := player1.eval()
	player2handType, player2highestCard := player2.eval()
	if player1handType > player2handType {
		return true
	} else if player1handType < player2handType {
		return false
	}
	return player1highestCard.rank > player2highestCard.rank // tie break
}
